packages/eslint-plugin/rules/no-top-level-mui4-imports.js

```diff
@@ -0,0 +1,129 @@
+      thirdLevelImport:
+        'Third level or deeper imports for Material UI are not allowed',
+    },
+    docs: {
+      description: 'Forbid top level import from Material UI v4 packages.',
+      url: 'https://github.com/backstage/backstage/blob/master/packages/eslint-plugin/docs/rules/no-top-level-mui4-imports.md',
+    },
+  },
+  create: context => ({
+    ImportDeclaration: node => {
+      // Return if empty import
+      if (node.specifiers.length === 0) return;
+      // Return if empty source value
+      if (!node.source.value) return;
+      // Return if source value not a string
+      if (typeof node.source.value !== 'string') return;
+      // Return if import does not start with '@material-ui/'
+      if (!node.source.value.startsWith('@material-ui/')) return;
+      // Return if import is from '@material-ui/core/styles', as it's valid already
+      if (node.source.value === '@material-ui/core/styles') return;
+      // Return if import is from '@material-ui/core/SvgIcon', as it's valid already
+      if (node.source.value === '@material-ui/core/SvgIcon') return;
+      // Return if proper import eg. `import Box from '@material-ui/core/Box'`
+      if (
+        node.specifiers.length === 1 &&
+        node.source.value?.split('/').length === 3
+      )
+        return;
+
+      // Report third level or deeper imports
+      if (
+        node.specifiers.length === 1 &&
+        node.source.value.split('/').length > 3
+      ) {
+        context.report({
+          node,
+          messageId: 'thirdLevelImport',
+        });
+        return;
+      }
+
+      // Report all other imports
+      context.report({
+        node,
+        messageId: 'topLevelImport',
+        fix: fixer => {
+          const replacements = [];
+          const styles = [];
+          const svgIcon = [];
+
+          const specifiers = node.specifiers.filter(
+            s => s.type === 'ImportSpecifier',
+          );
+
+          for (const specifier of specifiers) {
+            if (specifier.local.name === 'TabProps') {
+              replacements.push(
+                `import { TabProps } from '@material-ui/core/Tab';`,
+              );
+            } else if (
+              specifier.local.name === 'SvgIcon' ||
+              specifier.local.name === 'SvgIconProps'
+            ) {
+              svgIcon.push(specifier.local.name);
+            } else if (KNOWN_STYLES.includes(specifier.local.name)) {
+              styles.push(specifier.local.name);
+            } else {
+              const replacement = `import ${specifier.local.name} from '${node.source.value}/${specifier.local.name}';`;
+              replacements.push(replacement);
+            }
+          }
+
+          if (svgIcon.length > 0) {
+            if (svgIcon.every(s => ['SvgIcon', 'SvgIconProps'].includes(s))) {
+              replacements.push(
+                `import SvgIcon, { SvgIconProps } from '@material-ui/core/SvgIcon';`,
```

Comment:
Yeah I think this one speaks to my previous comment doesn't it, about merging several things on one line when they seem to properly refer to the same top-level import.

Maybe the code needs to be instead collecting everything into kind of a map, where the key is e.g. `SvgIcon` and then the value can say "emitComponent: boolean, emitProp: boolean" or whatnot, and then when you serialize that into lines you know whether to add the `Foo` and/or `{ FooProps }` bits and join them with a comma. Then the Tabs and SvgIcon special cases disappear

Reply:
Ok, applied your map suggestion and now there's nothing hard coded and it seems better to me. Might be some opportunities for improvements still so would welcome some more feedback